### ISSUE TITLE
[bitnami/etcd] Pods not recovering - Can't update initialClusterState

### ISSUE THREAD

**Issue by Lavaburn** [bitnami/charts#6251]
**Which chart**: bitnami/etcd 6.2.3

**Describe the bug**
Pods are crashing when changing initialClusterState from "new" to "existing".

```
etcd 10:44:26.03 INFO ==> Updating member in existing cluster
Error: bad member ID arg (strconv.ParseUint: parsing "": invalid syntax), expecting ID in Hex
```

**To Reproduce**
1. Deploy this chart with initialClusterState: "new" (extra values below)
2. Update this chart with new values: initialClusterState: "existing"

**Expected behavior**
When leaving the initialClusterState: "new" (or not setting it), pods crashing are not recovering. It is expected that setting initialClusterState to existing would resolve this.

When updating the initialClusterState to "existing", the pod should rejoin the cluster and be able to recover from pod crash.

**Version of Helm and Kubernetes**:

- Output of `helm version`:
```
version.BuildInfo{Version:"v3.5.3", GitCommit:"041ce5a2c17a58be0fcd5f5e16fb3e7e95fea622", GitTreeState:"dirty", GoVersion:"go1.15.8"}
```
- Output of `kubectl version`:
```
Client Version: version.Info{Major:"1", Minor:"20", GitVersion:"v1.20.5", GitCommit:"6b1d87acf3c8253c123756b9e61dac642678305f", GitTreeState:"clean", BuildDate:"2021-03-18T01:10:43Z", GoVersion:"go1.15.8", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"16", GitVersion:"v1.16.10", GitCommit:"f3add640dbcd4f3c33a7749f38baaac0b3fe810d", GitTreeState:"clean", BuildDate:"2020-05-20T13:51:56Z", GoVersion:"go1.13.9", Compiler:"gc", Platform:"linux/amd64"}
```

**Additional context**

values.yaml:
```
replicaCount: 3

auth:
  rbac:
    enabled: false

persistence:
  enabled: true
```

**Comment by jaspermarcus** [comment]
I believe we could be running into something similar in our deployments on helm chart version `6.2.0`. We periodically deploy changes via `helm upgrade --install`, and hence `ETCD_INITIAL_CLUSTER_STATE` also transitioned for us at some point from `new` to `existing`. 

While there don't seem to be issues during regular operations, we have noticed significant flakiness when a member is terminated for any reason (e.g node draining, OOM killed) and is subsequently unable to rejoin the cluster with the exact same error message

```
etcd 22:03:27.98 INFO  ==> Initializing etcd
etcd 22:03:27.98 INFO  ==> Detected data from previous deployments
etcd 22:03:38.17 INFO  ==> Updating member in existing cluster
Error: bad member ID arg (strconv.ParseUint: parsing "": invalid syntax), expecting ID in Hex
```

After following the conversation on a similar sounding issue (https://github.com/bitnami/charts/issues/3190), I did some debugging around the creation of the `${ETCD_DATA_DIR}/member_id` file. 

Could it ever be expected that the `member_id` file is created, but empty? 
```
I have no name!@storage-etcd-3:/bitnami/etcd/data$ ls -al
total 16
drwx--S--- 3 1001 1001 4096 Apr 26 12:39 .
drwxrwsr-x 4 root 1001 4096 Apr 26 12:37 ..
drwx--S--- 4 1001 1001 4096 Apr 26 12:39 member
-rwx------ 1 1001 1001    0 Apr 23 16:14 member_id
-rwx------ 1 1001 1001  611 Apr 23 16:14 new_member_envs
I have no name!@storage-etcd-3:/bitnami/etcd/data$ cat member_id
I have no name!@storage-etcd-3:/bitnami/etcd/data$
```
The above output is from `kubectl exec`-ing into a member in the cluster where we were observing this issue, and was consistent across all members of our 5 member cluster.

**Comment by alemorcuq** [comment]
Yes, I don't think the `member_id` being empty is ever expected, @jaspermarcus. That seems to be the issue here.

@Lavaburn can you confirm that file is also empty in your case?

I'm going to create an internal task to investigate this issue.

**Comment by Lavaburn** [comment]
I did some further investigation and was finally able to get the cluster up.
At the time that the issue was present, indeed, member_id was empty. This is certainly causing the error message (and unexpected behavior).

To resolve my issue, I had to clean the persistent volumes. I then ran into other issues with the livenessProbe, and the pod disruption not being set by default.
The current cluster is able to restore from failure, but there is no member_id file (not sure whether this is expected).

ls /bitnami/etcd/data/
member  new_member_envs

**Comment by alemorcuq** [comment]
Thanks for the information, @Lavaburn. I will add it to the task.

**Comment by DanJake** [comment]
Hello @alemorcuq, do you have any update about this issue

**Comment by alemorcuq** [comment]
Hi, @DanJake. No news so far.

**Comment by billylindeman** [comment]
I'm also experiencing this issue

**Comment by alemorcuq** [comment]
Thanks, @billylindeman. I have just raised the internal priority of this issue.

**Comment by aavandry** [comment]
Hi,
i have the same issue. I use bitnami/etcd 6.2.11. When cluster moves to existing state, the member_id file became empty, so a POD can not start. I use AWS Fargate nodes with EFS PV, EFS volumes are mounted to separate EC2, just for checking.

3 EFS volumes are mounted, but only 2 member_id files have data -
```
root@EfsEndpoint:/mnt# cat fs-*/data/member_id                                                                                                                                                                                                                                 
a11bd9d334c23662                                                                                                                                                                                                                                                               
cd512f3f4ef97259     
```
the 3rd file has zero size
```
root@EfsEndpoint:/mnt# ls -la fs-*/data/member_id                                                                                                                                                                                                                              
-rw-r--r-- 1 andry root  0 Jul 27 14:06 fs-2a126d9e_dev-data-etcd-1/data/member_id                                                                                                                                                                                             
-rw-r--r-- 1 andry root 17 Jul 27 14:06 fs-5c136ce8_dev-data-etcd-0/data/member_id                                                                                                                                                                                             
-rw-r--r-- 1 andry root 17 Jul 27 14:06 fs-6a126dde_dev-data-etcd-2/data/member_id 
```

**Comment by alemorcuq** [comment]
We will start working on this ASAP, @aavandry. Thanks for reporting.

**Comment by juv** [comment]
Is there any workaround for this? 
I spun up a fresh cluster with 5 pods, deleted two pods randomly to test HA capabilities and now all etcd pods are failing.

**Comment by juv** [comment]
@alemorcuq can you give a few hints on how to fix this issue? The current state of the helm chart is basically unusable as far as I can tell, as any pod crash will lead to a infinite crash loop or am I missing something?

**Comment by ckoehn** [comment]
@alemorcuq Is there a rough estimate when this will be worked on? We are experiencing the very same issue and hesitant to deploy the current state to any production environment.

**Comment by alemorcuq** [comment]
If everything goes according to plan, we will start working on this issue next week.

**Comment by christianhuening** [comment]
do we know in which version of the helm chart this issue was introduced? Is there an older version that works fine?

**Comment by hydramyst** [comment]
hello, i have same issue too.

**Comment by ckoehn** [comment]
> @alemorcuq can you give a few hints on how to fix this issue? The current state of the helm chart is basically unusable as far as I can tell, as any pod crash will lead to a infinite crash loop or am I missing something?

You might want to try https://github.com/bitnami/bitnami-docker-etcd/pull/21.

**Comment by juv** [comment]
@ckoehn thanks, I built your PR and pushed it to Dockerhub, just in case someone wants to try out without needing to build it yourself: https://hub.docker.com/layers/juvd/bitnami-docker-etcd/pr-21-3.5.0-debian-10-r61/images/sha256-d642961590041f0922a19a4f3137b82586eaf692ce82a8d3f29a0699231f7e76

I made sure to double-check whether your changes were really built into my image. They are.
`/opt/bitnami/scripts/etcd/prestop.sh` for example
```
I have no name!@chart-1629734060-etcd-1:/opt/bitnami/scripts/etcd$ cat prestop.sh 
#!/bin/bash

# shellcheck disable=SC1091

set -o errexit
set -o pipefail
set -o nounset
# set -o xtrace # Uncomment this line for debugging purposes

# Load libraries
. /opt/bitnami/scripts/libetcd.sh

# Load etcd environment settings
. /opt/bitnami/scripts/etcd-env.sh

endpoints="$(etcdctl_get_endpoints true)"
if is_empty_value "${endpoints}"; then
    exit 0
fi
read -r -a extra_flags <<< "$(etcdctl_auth_flags)"
extra_flags+=("--endpoints=${endpoints}" "--debug=true")
etcdctl member remove "$(cat "${ETCD_DATA_DIR}/member_id")" "${extra_flags[@]}" > "$(dirname "$ETCD_DATA_DIR")/member_removal.log"
```

The initial cluster setup after running `helm install` with a replicaCount of `3` seems to run fine. 
All three pods have a different `member_id`. 
After killing one of the etcd pods, I think it was the `-0` pod, this got into a crash loop:

```
etcd 16:04:14.83 
etcd 16:04:14.83 Welcome to the Bitnami etcd container
etcd 16:04:14.83 Subscribe to project updates by watching https://github.com/bitnami/bitnami-docker-etcd
etcd 16:04:14.83 Submit issues and feature requests at https://github.com/bitnami/bitnami-docker-etcd/issues
etcd 16:04:14.83 
etcd 16:04:14.83 INFO  ==> ** Starting etcd setup **
etcd 16:04:14.84 INFO  ==> Validating settings in ETCD_* env vars..
etcd 16:04:14.84 INFO  ==> Initializing etcd
etcd 16:04:14.84 INFO  ==> Detected data from previous deployments
etcd 16:04:15.12 INFO  ==> Updating member in existing cluster
{"level":"warn","ts":"2021-08-23T16:04:15.312Z","logger":"etcd-client","caller":"v3/retry_interceptor.go:62","msg":"retrying of unary invoker failed","target":"etcd-endpoints://0xc000372700/#initially=[chart-1629734060-etcd-1.chart-1629734060-etcd-headless.default.svc.cluster.local:2379;chart-1629734060-etcd-2.chart-1629734060-etcd-headless.default.svc.cluster.local:2379]","attempt":0,"error":"rpc error: code = NotFound desc = etcdserver: member not found"}
Error: etcdserver: member not found
```

While pod `-0` was still crashing with the above error, I tried to scaling down the entire statefulset to 0 with `kubectl scale sts {name} --replicas=0`, then back to 3. It had me ending up in a crashloopbackoff for 2 out of 3 pods.
You can find the logs attached for this scenario.

![image](https://user-images.githubusercontent.com/8269353/130481219-4de25a25-3757-4dd1-addc-7bb78cadf894.png)
[chart-1629734060-etcd-1.log](https://github.com/bitnami/charts/files/7033375/chart-1629734060-etcd-1.log)
[chart-1629734060-etcd-2.log](https://github.com/bitnami/charts/files/7033376/chart-1629734060-etcd-2.log)
[chart-1629734060-etcd-0.log](https://github.com/bitnami/charts/files/7033377/chart-1629734060-etcd-0.log)

I can't seem to be able to reproduce this with a replicaCount of `5`. All looks good here. I tested with the same procedure: randomly kill pods, wait for them to rejoin the cluster, kill one more, then wait again, scale down to 0 replicas, scale back up to 5 replicas. Everything went smooth.

Then I proceeded to test again with a replicaCount of `3`. While killing pods randomly, it all worked fine here.
Scaling to 0, back to 3 got the `-0` pod into a crashloop again. It eventually fixed itself after 4 retries though, which is good.
![image](https://user-images.githubusercontent.com/8269353/130484348-a64a20e7-7dce-40c6-af7a-1cf227d59e35.png)

To sum up, I might have got my cluster into a weird state when scaling down to 0 pods while one of the pods was still in a crashloop. Your fixes make the etcd deployment look _way_ better than in the current bitnami docker image!

**Comment by juan131** [comment]
Thanks for sharing your experience everyone! I'll try to review the PR created by @ckoehn today and hopefully we should have a solution for this race condition during the week.

**Comment by kuznero** [comment]
I also tried building docker image based on mentioned [PR](https://github.com/bitnami/bitnami-docker-etcd/pull/21) - it does not generate an event that says that pre-stop failed when scaling down, but it does not help re-joining the cluster when scaling up unfortunately.

**Comment by juan131** [comment]
Hi @kuznero 

Thanks for sharing the feedback, please note this is WIP and there are things we need to polish yet. One question: are you respecting the quorum when scaling down? I mean, how many replicas did you have before and after scaling down?

**Comment by kuznero** [comment]
Hi @juan131,

I had 3 initially, and then I scaled down to 2 (still majority), but granted the third node managed to send member removing signal to all the other nodes (1, 2) it should not matter, right? The resulting cluster will have 2 nodes only (not perfect from consensus perspective since next pod kill will result in a brain split situation).

**Comment by juan131** [comment]
Correct @kuznero ... Although 2 nodes it's not ideal for consensus, you should be able to scale back to 3 nodes if the "member removal" command worked as expected. Are you using `helm upgrade MY-RELEASE bitnami/etcd --set replicaCount=X` or `kubectl scale --replicas=X statefulset/MY-RELEASE-etcd` for scaling the solution?

**Comment by kuznero** [comment]
Hi @juan131, I am scaling in/out directly on a statefulset for testing purposes.

**Comment by ckoehn** [comment]
Keep in mind you might need to adjust INITIAL_CLUSTER_STATE too.

**Comment by kuznero** [comment]
Hi @ckoehn, 

Do you mean that I can try re-testing from the PR while trying to scale down/up through `helm update` instead of triggering it manually?

**Comment by juan131** [comment]
Hi @ckoehn @kuznero 

The `INITIAL_CLUSTER_STATE` shouldn't affect this kind of scaling operations and it should be possible to scale using `kubectl scale --replicas=X statefulset/MY-RELEASE-etcd`. If that's not the case, then this can be considered another bug to be fixed.

**Comment by danfromtitan** [comment]
Came across this issue while investigating pods crash looping. Let me start by saying there is not one single cause for this, there are rather a number of scenarios that could result in cluster instability.

Tested with juvd/bitnami-docker-etcd:pr-21-3.5.0-debian-10-r61 image, bitnami/bitnami-docker-etcd#21 almost fixes the missing member_id, but it won't address the issue with pods going CrashLoopBack during scaling.


Listed below how things can go wrong:

- When scaling down nodes, both `kubectl scale` and `helm upgrade` would leave the data PVCs behind. Unless a post upgrade hook gets added to the chart, you need to delete those yourself. That way when you scale-up, the nodes come with empty data instead of carrying artifacts from the previous run that could possibly mess-up the cluster state. The [libetcd.sh](https://raw.githubusercontent.com/bitnami/bitnami-docker-etcd/master/3/debian-10/rootfs/opt/bitnami/scripts/libetcd.sh) script covers for "Detected data from previous deployments", but you're left at the mercy of etcd's ability to cope with old data. I'd rather start new nodes fresh and sync the latest data.

- Scaling down with helm upgrade, I would expect the remaining nodes to be left unchanged as long as there is no image update. What I see is remaining nodes get re-created one at a time. That happens because there are container env var changes:

ETCD_INITIAL_CLUSTER_STATE changes from new -> existing
ETCD_INITIAL_CLUSTER changes from listing 5 node URLs to 3

Now, etcd makes use of these env vars _only_ to start a new member after that member has been added to the existing cluster. In other words existing cluster members are informed of the new addition via `etcdctl member add`, then it's _only_ the newly added member that should be made aware of, hey, you get added to an existing cluster and this is the list of node URLs including yours. IMO the choice to set these two env vars in running containers permanently was unfortunate and it's the most common reason why cluster becomes unstable (i.e. quorum gets messed-up). 


- Scaling down with `kubectl scale` won't re-create remaining nodes, but if you scale-up with kubectl that won't update ETCD_INITIAL_CLUSTER_STATE and ETCD_INITIAL_CLUSTER in statefulset's manifest, so the new nodes that come up won't have the proper ETCD_INITIAL_CLUSTER list (unless you scale-up to the same number of nodes you had initially)


- Scaling more than one node at a time might cause the cluster to become unstable, that's being discussed in https://github.com/etcd-io/etcd/issues/10738


Bottom line, there are one or two fixes needed in the Helm chart:

- move ETCD_INITIAL_CLUSTER_STATE and ETCD_INITIAL_CLUSTER outside container variables, instead handle these inside the libetcd.sh script. This would not only prevent remaining nodes to be re-created but it would also make it possible to scale properly using `kubectl scale`
- add a post-upgrade hook to delete PVCs left after node scale down (nice to have)

**Comment by juan131** [comment]
Hi @danfromtitan  

Thanks for the detailed analysis. Every single detail you mentioned is actually true!!

> When scaling down nodes, both `kubectl scale` and `helm upgrade` would leave the data PVCs behind. Unless a post upgrade hook gets added to the chart, you need to delete those yourself. That way when you scale-up, the nodes come with empty data instead of carrying artifacts from the previous run that could possibly mess-up the cluster state. The libetcd.sh script covers for "Detected data from previous deployments", but you're left at the mercy of etcd's ability to cope with old data. I'd rather start new nodes fresh and sync the latest data.

That's actually a very good point. The main problem here is that we can't distinguish (at container logic level) between upgrades and container restarts. I mean, I can't think of a way for the container (without consulting the K8s API) to know if the detected persisted data should be discarded or not.

You mentioned "post-upgrade" hooks but that's not a possibility we're willing to explore for a very simple reason: Bitnami charts are widely used and many users do NOT use **helm** to manage their apps. Instead, they render the templates to pure YAML manifests and install them using **kubectl**. Therefore, we can't rely on Helm hooks.

> move ETCD_INITIAL_CLUSTER_STATE and ETCD_INITIAL_CLUSTER outside container variables, instead handle these inside the libetcd.sh script. This would not only prevent remaining nodes to be re-created but it would also make it possible to scale properly using `kubectl scale`.

I like the idea, but it can be quite challenging to do so unless we provide the container the capabilities and RBAC permissions to talk with the K8s API to discover this information. Instead, we could improve the logic inside the libetcd.sh script to make it smarter and capable of distinguish when to ignore the values set on these env. vars.

**Comment by danfromtitan** [comment]
> I like the idea, but it can be quite challenging to do so unless we provide the container the capabilities and RBAC permissions to talk with the K8s API to discover this information.

There is no need for k8s API access to determine the list that goes in ETCD_INITIAL_CLUSTER, etcdctl would do:

```bash
etcdctl member list --user root:edited --cert /opt/bitnami/etcd/certs/client/cert.pem --key /opt/bitnami/etcd/certs/client/key.pem --cacert /opt/bitnami/etcd/certs/client/ca.crt --endpoints=etcd-headless.kv.svc.cluster.local:2379
f5859cbead4c581, started, etcd-1, https://etcd-1.etcd-headless.kv.svc.cluster.local:2380, https://etcd-1.etcd-headless.kv.svc.cluster.local:2379, false
1ac88be59a5bff17, started, etcd-3, https://etcd-3.etcd-headless.kv.svc.cluster.local:2380, https://etcd-3.etcd-headless.kv.svc.cluster.local:2379, false
421f43645e67ee1e, started, etcd-2, https://etcd-2.etcd-headless.kv.svc.cluster.local:2380, https://etcd-2.etcd-headless.kv.svc.cluster.local:2379, false
```

Using `etcd-headless.kv.svc.cluster.local` to drive the member list has the added benefit that etcdctl would retry by itself until it finds a node that is responsive (I'm thinking here of the situation when some of the nodes are not started or crash looping, still they resolve from the headless service DNS record).
Having now a reliable cluster trusted list of nodes, you can parse the output above to build the ETCD_INITIAL_CLUSTER list + add the new member hostname to it.

>  Instead, we could improve the logic inside the libetcd.sh script to make it smarter and capable of distinguish when to ignore the values set on these env. vars.

I'll continue on this in https://github.com/bitnami/bitnami-docker-etcd/pull/21, that seems like a better place for scripts discussion.

**Comment by juan131** [comment]
Yes please, let's continue the conversation in the PR. Thanks so much

**Comment by nilimaomni** [comment]
@juan131 Any update on the above mentioned? I'm still facing etcd pod crashloopbackoff issue.

**Comment by carrodher** [comment]
Thank you for reaching out and for your interest in our project. We'd be happy to assist you, but to do so effectively, we'll need some more details. Could you please create a separate issue filling the Issue template?

Providing this information will help us better understand your situation and assist you in a more targeted manner.
